Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,218 @@
+                                    self.configuration.loggingConfiguration.serviceKey: "\(service)",
+                                ]
+                            )
+
+                            group.cancelAll()
+                            throw ServiceRunnerError.serviceFinishedUnexpectedly()
+                        } else {
+                            // This service finished early but we expected it.
+                            // So we are just going to wait for the next service
+                            self.logger.debug(
+                                "Service finished",
+                                metadata: [
+                                    self.configuration.loggingConfiguration.serviceKey: "\(service)",
+                                ]
+                            )
+
+                            // We have to keep track of how many services finished to make sure
+                            // to stop when only the signal child task is left
+                            finishedServiceCount += 1
+                            if finishedServiceCount == self.services.count {
+                                // Every service finished. We can cancel the signal handling now
+                                // and return
+                                group.cancelAll()
+                                return
+                            } else {
+                                // There are still running services that we have to wait for
+                                continue
+                            }
+                        }
+
+                    case .serviceThrew(let service, let error):
+                        // One of the servers threw an error. We have to cancel everything else now.
+                        self.logger.error(
+                            "Service threw error. Cancelling all other services now",
+                            metadata: [
+                                self.configuration.loggingConfiguration.serviceKey: "\(service)",
+                                self.configuration.loggingConfiguration.errorKey: "\(error)",
+                            ]
+                        )
+                        group.cancelAll()
+
+                        throw error
+
+                    case .signalCaught(let unixSignal):
+                        // We got a signal. Let's initiate graceful shutdown in reverse order than we started the
+                        // services. This allows the users to declare a hierarchy with the order they passed
+                        // the services.
+                        self.logger.info(
+                            "Signal caught. Shutting down services",
+                            metadata: [
+                                self.configuration.loggingConfiguration.signalKey: "\(unixSignal)",
+                            ]
+                        )
+
+                        await shutdownGracefullyManager.shutdownGracefully()
```

Comment:
Do we want to allow this to run indefinitely?

Reply:
Why run `indefinitely`? The `wait` part is only there to hop to the underlying actor the handlers are all `sync`

Reply:
Ah, I missed that, I thought we were awaiting for the shutdown to complete.